### PR TITLE
Fix #13: add base scrapper for budget

### LINKED ISSUE #13: Create budget expenditure and receipts crawlers

### PR DESCRIPTION
**Pull request by shreyaagrawal0809**

### CODE REVIEW

**Review by curioswati — COMMENTED** [review]
please remove the `datasets` dir, specifically the `*ddo_codes` files from the PR as there's already an issue #8 open for this.

**Comment by shreyaagrawal0809** [comment]
Deleted please checkonce

File: scraper/scraper/spiders/budget_base.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        return self.make_dataset_request(params)
+
+    def parse_dataset(self, response):  #pylint: disable=no-self-use
+        '''
+        Parse each dataset page to collect the data in a csv file.
+        output: a csv file named with budget_expenditures_date format.
+        '''
+        # header row for the file.
+        heads = response.xpath('//table//tr[@class="popupheadingeKosh"]//td//text()').extract()
+        pdb.set_trace()
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_base.py]: remove pdb call

File: scraper/scraper/spiders/budget_base_pdf.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   #     file = open(pdf_names, 'wb')
+    #    file.write(response.body)
+     #   file.close()
+    
+    
+ 
+
+       
+       
+
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_base_pdf.py]: please remove these extra lines at the end of file

File: scraper/scraper/spiders/budget_base_pdf.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            file.write(response.body)
+            file.close()
+        else:
+            pdf_names = '../../datasets/receipts/' + response.meta['pdf_name_receipts'] + '.pdf'
+            file = open(pdf_names, 'wb')
+            file.write(response.body)
+            file.close()
+
+
+ #   def save_pdf_receipts(self, response):
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_base_pdf.py]: can we get rid of this function

**Review by curioswati — CHANGES_REQUESTED** [review]

File: scraper/scraper/spiders/budget_base.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                'from_date': params['date'],  # format: yyyymmdd
+                'To_date': params['date'],      # format: yyyymmdd
+                'ddlquery': params['query_id'],
+                'HODCode': params['hod_name'],
+                'Str': params['query_name'],
+                'Unit' : params['unit']
+            }
+
+            yield scrapy.Request(
+                self.query_url.format(urlencode(query_params)), # pylint: disable=no-member
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_base.py]: can we not disable this and resolve?

File: scraper/scraper/spiders/budget_base.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                and not hasattr(self, 'query_name')
+                and not hasattr(self, 'hod_code')
+            ):
+            yield scrapy.Request(self.start_urls[0], self.parse)
+        else:
+            for ddo_code in self.get_ddo_codes(self.treasury_id): #pylint: disable=no-member
+                params = {
+                    'date': self.date,  #pylint: disable=no-member
+                    'query_id': self.query_id,  #pylint: disable=no-member
+                    'query_name': self.query_name,  #pylint: disable=no-member
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_base.py]: can we get rid of these `disable=no-member`

File: scraper/scraper/spiders/budget_base.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        params = {
+            'date': self.date,  #pylint: disable=no-member
+            'query_id': query_id,
+            'query_name': query_name,
+            'hod_name': hod_name,
+            'unit': self.unit
+            }
+        return self.make_dataset_request(params)
+
+    def parse_dataset(self, response):  #pylint: disable=no-self-use
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_base.py]: you need not disable this, let the error show in pylint.

File: scraper/scraper/spiders/budget_exp_receipts.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    name = 'budget_expenditures'
+
+    start_urls = ['http://himkosh.hp.nic.in/treasuryportal/eKosh/ekoshhodAllocationquery.asp']
+
+    # dataset is collected from here.
+    query_url = 'http://himkosh.hp.nic.in/treasuryportal/eKosh/eKoshHODAllocationPopUp.asp?{}'
+
+    query_index = 9
+
+    unit = '.00001'
```
Comment (curioswati) [inline scraper/scraper/spiders/budget_exp_receipts.py:15]: fix this new line error

**Review by curioswati — CHANGES_REQUESTED** [review]

**Review by curioswati — APPROVED** [review]
good work!

**Pull request merged**